src/Orleans/Utils/IKeyedServiceCollection.cs

```diff
@@ -10,7 +12,40 @@ namespace Orleans.Runtime
     public interface IKeyedServiceCollection<in TKey, out TService>
         where TService : class
     {
-        TService GetService(TKey key);
+        TService GetService(IServiceProvider services, TKey key);
+    }
+
+    public interface IKeyedService<TKey, out TService> : IEquatable<TKey>
+    {
+        TService GetService(IServiceProvider services);
+    }
+
+    public class KeyedService<TKey, TService, TInstance> : IKeyedService<TKey, TService>
+            where TInstance : TService
+    {
+        private readonly TKey key;
+
+        public KeyedService(TKey key)
+        {
+            this.key = key;
+        }
+
+        public TService GetService(IServiceProvider services) => services.GetService<TInstance>();
```

Comment:
Ah, you are absolutely right about the simplicity of usage of the factories if injecting the dependency as a scoped service. That's a good enough reason to keep that as is at least for now.
BTW, you mentioned that factories should be scoped, but you are registering them as transient. Is this a mistake or by design?